### PR TITLE
spdx/parse: allow lowercase operators

### PR DESCRIPTION
**Pull request by soumyaDghosh**
Fixes https://forum.snapcraft.io/t/spdx-license-specification-allows-lower-case-expressions-but-we-are-getting-rejected-on-uploads/41296

----
Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

File: spdx/parser.go [hunk 1]
```diff
@@ -98,22 +101,22 @@ func (p *parser) validate(depth int) error {
 			if last == "" {
 				return fmt.Errorf("missing license before %s", tok)
 			}
-			if last == opAND || last == opOR {
+			if strings.ToUpper(last) == opAND || strings.ToUpper(last) == opOR {
```
Comment (bboozzoo) [inline spdx/parser.go:104]:
```suggestion
			if maybeLastOp := asOperator(last); maybeLastOp == opAND || maybeLastOp == opOR {
```

File: spdx/parser.go [hunk 2]
```diff
@@ -98,22 +101,22 @@ func (p *parser) validate(depth int) error {
...
 				return fmt.Errorf("expected license name, got %q", tok)
 			}
-			if tok == opWITH && last == "(" {
+			if upperTok == opWITH && last == "(" {
```
Comment (bboozzoo) [inline spdx/parser.go:107]:
```suggestion
			if asOperator(tok) == opWITH && last == "(" {
```

File: spdx/parser.go [hunk 3]
```diff
@@ -98,22 +101,22 @@ func (p *parser) validate(depth int) error {
...
 				return fmt.Errorf("expected license name before %s", tok)
 			}
-			if last == opWITH {
+			if strings.ToUpper(last) == opWITH {
```
Comment (bboozzoo) [inline spdx/parser.go:110]:
```suggestion
			if asOperator(last) == opWITH {
```

File: spdx/parser.go [hunk 4]
```diff
@@ -74,10 +76,11 @@ func (p *parser) validate(depth int) error {
 
 	for p.s.Scan() {
 		tok := p.s.Text()
+		upperTok := strings.ToUpper(tok)
```
Comment (bboozzoo) [inline spdx/parser.go:79]:
```suggestion
```

File: spdx/parser.go [hunk 5]
```diff
@@ -98,22 +101,22 @@ func (p *parser) validate(depth int) error {
...
 				return fmt.Errorf("expected license name before %s", tok)
 			}
-			if last == opWITH {
+			if strings.ToUpper(last) == opWITH {
 				return fmt.Errorf("expected exception name, got %q", tok)
 			}
 		default:
 			switch {
-			case last == opWITH:
+			case strings.ToUpper(last) == opWITH:
```
Comment (bboozzoo) [inline spdx/parser.go:115]:
```suggestion
			case asOperator(last) == opWITH:
```

File: spdx/parser.go [hunk 6]
```diff
@@ -98,22 +101,22 @@ func (p *parser) validate(depth int) error {
...
 				if _, err := newLicenseExceptionID(tok); err != nil {
 					return err
 				}
-			case last == "", last == opAND, last == opOR:
+			case last == "", isOperator(last):
```
Comment (bboozzoo) [inline spdx/parser.go:119]:
```suggestion
			case isOperator(last):
				// last token was an operator, expecting a license,
				// WITH already handled earlier
				fallthrough
			case last == "":
```

**Comment by soumyaDghosh** [comment]
Hi @bboozzoo your suggestions didn't work as expected. I'll try to take some more time to look into it. Very sorry for the delay. I'm thinking to just implement a helper function and replace the `strings.toUpper` instances with the helper function.

**Pull request closed without merging**